File: webapp/codigos-fonte/fila.c
#include <stdio.h>
#include <stdlib.h>

struct no {
    int valor;
    struct no *proximo;
};

typedef struct no No;

struct fila{
    No *cabeca;
    No *cauda;
};

typedef struct fila Fila;

void enfileirar(Fila *f, int valor) {
    No *novoNo = (No *) malloc(sizeof(No));
    novoNo->valor = valor;
    novoNo->proximo = NULL;

    if(f->cabeca == f->cauda && f->cabeca == NULL) {
        f->cabeca = f->cauda = novoNo;
    }
    else{
        f->cauda->proximo = novoNo;
        f->cauda = novoNo;
    }
}

void desenfileirar(Fila *f){
    if(f->cabeca == f->cauda && f->cabeca == NULL){
        printf("Erro ao tentar retirar um elemento de uma fila vazia\n");
        return;
    }
    No *temporario = f->cabeca;
    if(f->cabeca == f->cauda){
        f->cabeca = f->cauda = NULL;
    }
    else{
        f->cabeca = f->cabeca->proximo;
    }
    int valor = temporario->valor;
    free(temporario);
}

void printAll(Fila *f){
    No *atual = f->cabeca;
    while(atual != NULL){
        printf("%d ", atual->valor);
        atual = atual->proximo;
    }
    printf("\n");
}

Fila *criarFila(){
    Fila *f = (Fila *) malloc(sizeof(Fila));
    f->cabeca = f->cauda = NULL;
    return f;
}

int main() {
    Fila *f = criarFila();
    for(int i = 0; i < 10; i++){
        enfileirar(f, i);
        printAll(f);
    }
    for(int i = 0; i < 10; i++){
        desenfileirar(f);
        printAll(f);
    }
}

File: webapp/codigos-fonte/fila.py
class No:
    def __init__(self, valor, proximo):
        self.valor = valor
        self.proximo = proximo


class Fila:
    def __init__(self):
        self.cabeca = None
        self.cauda = None

    def enfileirar(self, valor):
        novo = No(valor, None)

        if self.cabeca == self.cauda and self.cabeca == None:
            self.cabeca = self.cauda = novo
        else:
            self.cauda.proximo = novo
            self.cauda = novo
    
    def desenfileirar(self):
        if self.cabeca == self.cauda and self.cabeca == None:
            print("Não pode retirar um elemento de uma fila vazia")
            return

        temporario = self.cabeca
        if self.cabeca == self.cauda:
            self.cabeca = self.cauda = None
        else:
            self.cabeca = self.cabeca.proximo
        
        valor = temporario.valor        


def printAll(fila):
    atual = fila.cabeca
    while(atual != None):
        print(atual.valor, end=" ")
        atual = atual.proximo
    print()


f = Fila()
for i in range(10):
    f.enfileirar(i)
    printAll(f)

for i in range(10):
    printAll(f)
    f.desenfileirar()